### PR TITLE
refactor fetch

### PR DESCRIPTION
**Pull request by tee8z**
Makes the fetch paginated based on time and defaults to grabbing events that have occurred in the last week. Also moves the shared logic between the vote image and tallying into a share function.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -331,143 +331,149 @@ <h2>Leaderboard Results</h2>
...
+                let events = [];
+                let eventids = [];
+                let completedRelays = 0;
+                const totalRelays = readRelays.length;
+
+                readRelays.forEach((relay) => {
+                    const ws = new WebSocket(relay);
+
+                    ws.onopen = () => {
+                        console.log(`Opening connection to ${relay} with until=${until}`);
```
Comment (vicariousdrama) [inline index.html]: if (window.DEBUG)

**Comment by tee8z** [comment]
@vicariousdrama 
```
main branch
currentResults: 
Array(716) [ {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, … ]

new logic:
Total events collected: 716
```

File: index.html [hunk 2]
```diff
@@ -331,143 +331,149 @@ <h2>Leaderboard Results</h2>
...
+                                if (hasNamespace && hasLabel && hasHashtag) {
+                                    events.push(nostrEvent);
+                                    eventids.push(nostrEvent.id);
                                 }
                             }
-                            if (message[0] === "EOSE") {
-                                ws.send(JSON.stringify(["CLOSE", "options"]));
-                                ws.close();
-                            }
-                            if (message[0] == "NOTICE") {
```
Comment (vicariousdrama) [inline index.html:379]: Why remove the NOTICE handler?

File: index.html [hunk 3]
```diff
@@ -331,143 +331,149 @@ <h2>Leaderboard Results</h2>
...
-                                if (message.length > 1) console.log(`notice from ${relay}: ${JSON.stringif(message[1])}`);
+                        }
+
+                        if (message[0] === "EOSE") {
+                            console.log(`Received EOSE from ${relay}`);
+                            completedRelays++;
+                            if (completedRelays === totalRelays) {
```
Comment (vicariousdrama) [inline index.html:394]: 👍🏽

File: index.html [hunk 4]
```diff
@@ -331,143 +331,149 @@ <h2>Leaderboard Results</h2>
...
 
-            async function fetchResults() {
-                return new Promise((resolve, reject) => {
-                    let results = [];
-                    let resultids = [];
+                const handleCompletion = () => {
+                    // Sort by timestamp
+                    events.sort((a, b) => b.created_at - a.created_at);
+
+                    console.log(`Total events collected: ${events.length}`);
```
Comment (vicariousdrama) [inline index.html]: if (window.DEBUG)

File: index.html (hunk 2 above)
Comment (tee8z) [reply index.html:379]: no reason, will add it back

**Comment by vicariousdrama** [comment]
If the intention is to get all events, paginated, then need to have iterative calls after EOSE that 1) closes subscription, 2) sets up next subscription with new since/until range, 3) continues going back covering rational period of time.  

We may want to consider a voting time frame (voting start/ends) tracked in a voting config record for future enhancement, but don't need to accomodate that specifically in this PR

**Comment by vicariousdrama** [comment]
> @vicariousdrama
> 
> ```
> main branch
> currentResults: 
> Array(716) [ {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, … ]
> 
> new logic:
> Total events collected: 716
> ```

I'm missing something logically as Im not seeing how the pagination is actuallly being performed.  You have the call changed to a pageSize of 500. This implies that it would be limited to that, or at most 500 * number of relays being interrogated. Prior to the change, the limit was 5000 which would explain how we were getting 716 results back.

**Comment by tee8z** [comment]
I really like that idea to set the time range to look at for votes and pepes! 

Also, within a week based on this logic, unless we re-upload the images we wont be able to see them to vote. I think forces us into needing to use the "lists" nip though for the image urls, then we could make that part of the config too, where we set the id of that nostr event that holds all the urls https://github.com/nostr-protocol/nips/blob/master/51.md

**Comment by vicariousdrama** [comment]
> I really like that idea to set the time range to look at for votes and pepes!
> 
> Also, within a week based on this logic, unless we re-upload the images we wont be able to see them to vote. I think forces us into needing to use the "lists" nip though for the image urls, then we could make that part of the config too, where we set the id of that nostr event that holds all the urls https://github.com/nostr-protocol/nips/blob/master/51.md

I'd like to see more use of kind 30388 https://cornychat.com/datatypes#kind30388slideset referenced in a new config (which itself would be a replaceable)

**Review by vicariousdrama — APPROVED** [review]

**Pull request merged**